### PR TITLE
fix: Add previous year to analytics partition

### PR DESCRIPTION
**Pull request by luciano-fiandesio**
- DHIS2-5990
- DHIS2-8144

- When an aggregate analytics query is run, certain Period Types may have the
first week of the period starting **in the previous year** of the given
Period Type. For instance, the Quarterly period starting on 2019/01/01
and ending in 2019/03/31 has the first week of the period starting on
2018/12/31.
Therefore, the system must take in consideration (add to the sql
_where_ condition) the previous year, otherwise a data mismatch is present when
comparing Weekly period types and Quarterly/Yearly/SixYearly and Monthly
period types.

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/table/PartitionUtils.java [hunk 1]
```diff
@@ -251,11 +260,26 @@ public static AnalyticsTablePartition getLatestTablePartition( List<AnalyticsTab
...
     }
+
+    /**
+     * Returns the year of the first week of this period
+     *
+     * @param period a {@see Period}
+     *
+     * @return the 4 digits representation of an year
+     */
+    private static int getYearOfFirstWeekFromPeriod( Period period )
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/table/PartitionUtils.java]: Are we taking the various weekly period types (starting Tuesday, Wednesday, etc) here?

**Comment by larshelge** [comment]
I am actually surprised this is needed. The thinking behind the current solution is that a week belongs to the month/quarter for which 4 days or more belongs. This is commonly accepted it seems. So what we do for those weeks that span two years and partitions is to put the period into the partition it will belong. E.g. if a week has 4 days belonging the 2019, the weekly values will be put in the 2019 partition. If a week has 4 days belonging to 2018, the weekly values will be put in the 2018 partition, etc.

Since it is only weekly and bi-weekly period type which span multiple years/partitions, I think this should handle the problem. Where am I wrong? :-)

**Comment by luciano-fiandesio** [comment]
> I am actually surprised this is needed. The thinking behind the current solution is that a week belongs to the month/quarter for which 4 days or more belongs. This is commonly accepted it seems. So what we do for those weeks that span two years and partitions is to put the period into the partition it will belong. E.g. if a week has 4 days belonging the 2019, the weekly values will be put in the 2019 partition. If a week has 4 days belonging to 2018, the weekly values will be put in the 2018 partition, etc.
> 
> Since it is only weekly and bi-weekly period type which span multiple years/partitions, I think this should handle the problem. Where am I wrong? :-)

Yes, I see your point. The problem (for the people who reported the bug) is that they are "confused" by the fact that say, a quarterly period yields different results if you use a weekly type for the same time range - in terms of aggregated data.
A weekly ISO period (starting on Monday) that starts in the previous year, is not computed when using weekly period as dimension.

**Comment by larshelge** [comment]
Yep agree. Given that week boundaries do not match month/quarter boundaries perfectly, and we do not allow dis-aggregation of aggregate data values (assign parts of a value to a period), this will never be perfect and I think the 4-day rule is the best we can do. It seems something weird is going on from that Jira issue though so worth while to check out.

**Pull request closed without merging**